node/subsystem-util/src/metered/mod.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of mpsc channels to be able to extract metrics.
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use futures::{channel::mpsc, task::Poll, task::Context, sink::SinkExt, stream::Stream};
+
+use std::result;
+use std::sync::Arc;
+use std::pin::Pin;
+
+mod bounded;
+mod unbounded;
+
+pub use self::bounded::*;
+pub use self::unbounded::*;
+
+/// A peek into the inner state of a meter.
+#[derive(Debug, Clone, Default)]
+pub struct Meter {
+	/// Name of the receiver and sender pair.
+	name: &'static str,
+	// fill state of the channel
+	fill: Arc<AtomicUsize>,
+}
+
+impl Meter {
+	/// Count the number of items queued up inside the channel.
+	pub fn queue_count(&self) -> usize {
+		// when obtaining we don't care much about off by one
+		// accuracy
+		self.fill.load(Ordering::Relaxed)
+	}
+
+	/// Obtain the name of the channel `Sender` and `Receiver` pair.
+	pub fn name(&self) -> &'static str {
```

Comment:
how names are going to be used? given the `snake_case` convention looks like somehow in prometheus, but that will like create a [cardinality issue](https://www.robustperception.io/cardinality-is-key)

Reply:
Currently they are not used as such, the idea was to provide a unique identifier that _could_ be used for proemtheus i.e. - currently it is mere convenience and the metrics for prometheus must be implemented deliberately.